feat: Nutrition facts: check values before sending them #4446

LorenzoMascia opened the pull request:
Changes:
	modified:   packages/smooth_app/lib/pages/product/nutrition_page/nutrition_page.dart
	new file:   packages/smooth_app/lib/pages/product/nutrition_page/nutrition_validator.dart
	modified:   packages/smooth_app/lib/pages/product/nutrition_page/widgets/nutrition_facts_editor.dart
	modified:   packages/smooth_app/lib/pages/text_field_helper.dart

### What
Created a class to perform nutrient validation for each product. The NutritionValidator class normalizes the input values based on the weight of the Serving Size, converting them into values relative to 100g or 100ml. The normalized values are then compared to the maximum values typically found in food products. The values are verified and displayed in red when the save button is pressed.

### Screenshot
![Screenshot 2025-04-24 153859](https://github.com/user-attachments/assets/2af666c5-5717-496b-b3f3-66d1ca1c60ce)


### Fixes bug(s)
Enhancement: Nutrition facts: check values before sending them #4446

### Part of

Commit:
skipped ix #4681: Move Hunger Games to a real view instead of overlay

Committer: Lorenzo Mascia <lorenzo.mascia@ericsson.com>

Changes :
	modified:   packages/smooth_app/lib/pages/hunger_games/question_card.dart
	modified:   packages/smooth_app/lib/pages/hunger_games/question_page.dart
	modified:   packages/smooth_app/lib/pages/preferences/user_preferences_contribute.dart
	modified:   packages/smooth_app/lib/pages/product/product_questions_widget.dart

Commit:
Fixes after code review

Committer: Lorenzo Mascia <lorenzo.mascia@ericsson.com>

Changes:
	modified:   packages/smooth_app/lib/pages/hunger_games/question_card.dart
	modified:   packages/smooth_app/lib/pages/hunger_games/question_image_thumbnail.dart
	modified:   packages/smooth_app/lib/pages/hunger_games/question_page.dart
	modified:   packages/smooth_app/lib/pages/preferences/user_preferences_contribute.dart
	modified:   packages/smooth_app/lib/pages/product/product_questions_widget.dart

Commit:
Fixes after second review

Committer: Lorenzo Mascia <lorenzo.mascia@ericsson.com>

Changes:
	modified:   packages/smooth_app/lib/pages/hunger_games/question_image_thumbnail.dart
	modified:   packages/smooth_app/lib/pages/hunger_games/question_page.dart
	modified:   packages/smooth_app/lib/pages/preferences/user_preferences_contribute.dart
	modified:   packages/smooth_app/lib/pages/product/product_questions_widget.dart

LorenzoMascia committed:
Update packages/smooth_app/lib/pages/hunger_games/question_card.dart

Co-authored-by: monsieurtanuki <fabrice_fontaine@hotmail.com>

Commit:
Fixes after third review

Committer: Lorenzo Mascia <lorenzo.mascia@ericsson.com>

Changes:
	modified:   lib/pages/hunger_games/question_image_thumbnail.dart
	modified:   lib/pages/hunger_games/question_page.dart

LorenzoMascia committed:
feat:Nutrition facts: check values before sending them #4446

Changes:
	modified:   packages/smooth_app/lib/pages/product/nutrition_page/nutrition_page.dart
	new file:   packages/smooth_app/lib/pages/product/nutrition_page/nutrition_validator.dart
	modified:   packages/smooth_app/lib/pages/product/nutrition_page/widgets/nutrition_facts_editor.dart
	modified:   packages/smooth_app/lib/pages/text_field_helper.dart

monsieurtanuki commented on packages/smooth_app/lib/pages/product/nutrition_page/nutrition_validator.dart:
```diff
@@ -0,0 +1,482 @@
+      return true;
+    }
+
+    switch (nutrient) {
+      case Nutrient.addedSugars:
+        return _validateAddedSugars(quantity, unit, servingSize);
```
Not very elegant.
An extension on `Nutrient` would be much better.

LorenzoMascia replied:
I refactored the code using a Map. Please, let me know if in this way it's better.

monsieurtanuki commented on packages/smooth_app/lib/pages/product/nutrition_page/nutrition_validator.dart:
```diff
@@ -0,0 +1,482 @@
+    }
+  }
+
+  bool _validateAddedSugars(double q, Unit u, double s) =>
```
We can afford to use words instead of letters for variables.
A computer wouldn't make the difference. A reviewer and a developer would.

LorenzoMascia replied:
Thank you, I appreciate the suggestion.

monsieurtanuki commented on packages/smooth_app/lib/pages/product/nutrition_page/nutrition_validator.dart:
```diff
@@ -0,0 +1,482 @@
+  }
+
+  bool _validateAddedSugars(double q, Unit u, double s) =>
+      _validateMax(_normalize(q, u, Unit.G, s), 50.0);
```
I have no idea where you picked the min/max values.

LorenzoMascia replied:
I obtained the values as the maximum possible values in foods. Some must simply be less than the weight, such as carbohydrate or sugar content, but in other cases, like zinc or calcium, it's not possible to simply assume a content less to the weight. The check prevents also potential errors in the use of measurement units, like 50g instead of 50mg

monsieurtanuki commented on packages/smooth_app/lib/pages/product/nutrition_page/nutrition_validator.dart:
```diff
@@ -0,0 +1,482 @@
+    }
+    // Energy conversions
+    if (from == Unit.KCAL && to == Unit.KJ) {
+      return 4.184;
```
Please use `NutrimentsHelper.fromKCalToKJ` and `NutrimentsHelper.fromKJtoKCal`.

LorenzoMascia replied:
Not sure on how to use them, This method does not converts values but returns conversion factors.

monsieurtanuki commented on packages/smooth_app/lib/pages/product/nutrition_page/nutrition_validator.dart:
```diff
@@ -0,0 +1,482 @@
+    return 1.0;
+  }
+
+  double? _parseDouble(String? input) {
```
I'm sure there are other places in the code where we parse a double. Please don't reinvent the wheel.

LorenzoMascia replied:
I changed the name of this method because it does not simply parse a double but it also reads the unit specified in the input string

monsieurtanuki commented on packages/smooth_app/lib/pages/product/nutrition_page/nutrition_page.dart:
```diff
@@ -249,6 +250,41 @@ class _NutritionPageLoadedState extends State<NutritionPageLoaded>
 
+    final NutritionValidator validator = NutritionValidator();
+
+    String servingSize = '100.0';
+
+    if (_nutritionContainer.perSize == PerSize.serving) {
+      servingSize = _nutritionContainer.servingSize;
+    }
```
Not sure what you want to do here, but a cleaner syntax would be:
```suggestion
    final String servingSize;
    if (_nutritionContainer.perSize == PerSize.serving) {
      servingSize = _nutritionContainer.servingSize;
    } else {
      servingSize = '100.0';
    }
```

LorenzoMascia replied:
The scope is to obtain the serving size weight. There are 2 options:  per serving and per 100 g/ml

![Screenshot 2025-04-27 191707](https://github.com/user-attachments/assets/baa24536-77d9-487f-95bb-227eee30ee88)

![Screenshot 2025-04-27 191722](https://github.com/user-attachments/assets/b9c553ce-f27a-41f5-b8c4-bece3ad8d4f9)

In the first case the weight is specified in the serving size field (and the user can specify the unit).

monsieurtanuki commented on packages/smooth_app/lib/pages/product/nutrition_page/nutrition_page.dart:
```diff
@@ -249,6 +250,41 @@ class _NutritionPageLoadedState extends State<NutritionPageLoaded>
+      final TextEditingControllerWithHistory controller =
+          _controllers[nutrient]!;
+      final String quantity = controller.text.trim();
+      final Unit u = _nutritionContainer.getUnit(nutrient);
```
PLEASE USE WORDS!
```suggestion
      final Unit unit = _nutritionContainer.getUnit(nutrient);
```

LorenzoMascia replied:
Got it

monsieurtanuki commented on packages/smooth_app/lib/pages/product/nutrition_page/nutrition_validator.dart:
```diff
@@ -0,0 +1,482 @@
+      return 1.0;
+    }
+    // Mass conversions
+    if (from == Unit.G && to == Unit.MILLI_G) {
```
I'm sure there are conversion methods somewhere else in the app.
Please use them.

LorenzoMascia replied:
I did the search looking for all occurrences of the Unit enum. I didn't find a similar method.

monsieurtanuki reviewed (commented):
Hi @LorenzoMascia!
Please have a look at my comments.
Remember also that I'm expecting a screenshot in #6555.

LorenzoMascia committed:
Fixes after first review

Changes:
	modified:   packages/smooth_app/lib/pages/product/nutrition_page/nutrition_page.dart
	modified:   packages/smooth_app/lib/pages/product/nutrition_page/nutrition_validator.dart

monsieurtanuki commented:
Hi @LorenzoMascia!
Conflicts to be resolved. More generally speaking, I suggest that your code format matches the project format (basically, comma endings).

teolemon commented:
/gemini review

LorenzoMascia committed:
Aligned question_image_thumbnail.dart with upstream/develop

LorenzoMascia committed:
Changes after Geminy review

	modified:   packages/smooth_app/lib/pages/product/nutrition_page/nutrition_page.dart
	modified:   packages/smooth_app/lib/pages/product/nutrition_page/nutrition_validator.dart

LorenzoMascia commented:
I’ve rebased the branch onto the latest develop to resolve all conflicts, and I've applied gemini's suggestions.

LorenzoMascia commented:
/gemini review

AshAman999 commented:
Hey @LorenzoMascia 
Any update on this ?
